io_scene_niftools/modules/nif_import/animation/transform.py

```diff
@@ -228,52 +258,70 @@ def import_keyframe_controller(self, n_kfc, b_obj, bone_name=None, n_bone_bind_s
             NifLog.debug('Rotation keys...(quaternions)')
             fcurves = self.create_fcurves(b_action, "rotation_quaternion", range(4), flags, bone_name)
             for t, val in rotations:
                 key = mathutils.Quaternion([val.w, val.x, val.y, val.z])
                 if bone_name:
-                    key = math.import_keymat(n_bone_bind_rot_inv, key.to_matrix().to_4x4()).to_quaternion()
+                    key = math.import_keymat(n_bind_rot_inv, key.to_matrix().to_4x4()).to_quaternion()
                 self.add_key(fcurves, t, key, interp_rot)
         if translations:
             NifLog.debug('Translation keys...')
             fcurves = self.create_fcurves(b_action, "location", range(3), flags, bone_name)
             for t, val in translations:
                 key = mathutils.Vector([val.x, val.y, val.z])
                 if bone_name:
-                    key = math.import_keymat(n_bone_bind_rot_inv, mathutils.Matrix.Translation(key - n_bone_bind_trans)).to_translation()
+                    key = math.import_keymat(n_bind_rot_inv, mathutils.Matrix.Translation(key - n_bind_trans)).to_translation()
                 self.add_key(fcurves, t, key, interp_loc)
         if scales:
             NifLog.debug('Scale keys...')
             fcurves = self.create_fcurves(b_action, "scale", range(3), flags, bone_name)
             for t, val in scales:
                 key = (val, val, val)
                 self.add_key(fcurves, t, key, interp_scale)
+        return b_action
 
     def import_transforms(self, n_block, b_obj, bone_name=None):
         """Loads an animation attached to a nif block."""
         # find keyframe controller
-        n_kfc = math.find_controller(n_block, NifFormat.NiKeyframeController)
-        # try again
-        if not n_kfc:
-            n_kfc = math.find_controller(n_block, NifFormat.NiTransformController)
+        n_kfc = math.find_controller(n_block, (NifFormat.NiKeyframeController, NifFormat.NiTransformController))
         if n_kfc:
             # skeletal animation
             if bone_name:
-                bone_bm = math.import_matrix(n_block)  # base pose
-                n_bone_bind_scale, n_bone_bind_rot, n_bone_bind_trans = math.decompose_srt(bone_bm)
-                self.import_keyframe_controller(n_kfc, b_obj, bone_name, n_bone_bind_scale, n_bone_bind_rot.inverted(), n_bone_bind_trans)
+                p_bone = b_obj.pose.bones[bone_name]
+                self.import_keyframe_controller(n_kfc, b_obj, p_bone, f"{b_obj.name}_Anim")
             # object-level animation
             else:
-                self.create_action(b_obj, b_obj.name + "-Anim")
-                self.import_keyframe_controller(n_kfc, b_obj)
+                self.import_keyframe_controller(n_kfc, None, b_obj, f"{b_obj.name}_Anim")
+
+    def import_controller_manager(self, n_block, b_obj, b_armature):
+        ctrlm = n_block.controller
+        if ctrlm and isinstance(ctrlm, NifFormat.NiControllerManager):
```

Comment:
Is this only every going to be on the root node?

Reply:
It doesn't matter which node this is on for the code to work, could technically also be on other nodes.

Reply:
Think this is fine for now, this is just part of the limitation of having to know ahead of time, vs processing on demand. 
Good enough for now.